UIE-165 error.ts more decoupling for Environments code sharing

msilva-broad opened the pull request:
### Jira Ticket: https://broadworkbench.atlassian.net/browse/UIE-165

<!-- ### Dependencies --->
<!-- Include any dependent tickets and describe the relationship. Include any relevant Jira tickets. --->

## Summary of changes:
Complete the dependency decoupling on error.ts utils to be workable for Environments component code sharing goals for AoU.

### What
- promoted error.ts to notifications package.
- removed dependencies to terra-ui specific things like session timeout ignore-error check with generic IgnoreErrorDecider function callback.
- doing this entangled the use of withErrorReporter to require an additional arg on each use... so instead, this was changed to makeNotificationsProvider, with the NotificationsProvider placed in react context with useNotificationsFromContext hook being shifted and expanded to directly expose a notify method and the withErrorReporter methods.
- composing apps are encouraged to leverage makeNotificationsProvider to initialize the value of useNotificationsFromContext context provider at the root of app visual tree.
- withErrorReporting variant methods like withErrorReporingInModal have been consolidated in the NotificationsProvider interface to all come from a parameterized withErrorReporting method:
  - i.e. withErrorReportingInModal becomes:
     `withErrorReporting('Error deleting cloud environment', { rethrow: true, onReported: onDismiss })`
- backwards compatible proxies for error methods are still in place, to leave code not ready for code-sharing unchanged.

### Why
- unblock eventual cross-app code sharing of Environments and other complex components.

### Testing strategy
<!-- Note that changes impacting components in Storybook stories can be viewed at
https://www.chromatic.com/library?appId=65fc89c9335768720ff8605a&branch=<branch>
The branch corresponding to this PR is selected, and changes can be reviewed by commit. --->

- [ ] No functional change expected.  Unit test coverage is robust.

<!-- ### Visual Aids -->
<!-- https://support.apple.com/guide/quicktime-player/record-your-screen-qtp97b08e666/mac -->

msilva-broad committed:
UIE-165 error.ts more decoupling for Environments code sharing
- promoted error.ts to notifications package.
- removed dependencies to terra-ui specific things like session timeout ignore-error check with generic IgnoreErrorDecider function callback.
- doing this entangled the use of withErrorReporter to require an additional arg on each use... so instead, this was changed to makeNotificationsContract, with the NotificationsContract placed in react context with useNotifications hook being shifted and expanded to directly expose a notify method and the withErrorReporter methods.
- composing apps are encouraged to leverage makeNotificationsContract to initialize the value of useNotifications context provider at the root of app visual tree.

msilva-broad committed:
UIE-165 error.ts more decoupling for Environments code sharing
- fixed Environments storybook setup to match new Notifications setup

msilva-broad committed:
UIE-165 error.ts more decoupling for Environments code sharing
- fixed a type error in BillingList.ts that recent type improvements exposed through known issues with _.flow type flow.

nawatts commented on packages/notifications/package.json:
```diff
@@ -24,6 +24,7 @@
   ],
   "dependencies": {
     "@terra-ui-packages/components": "^0.0.6",
+    "@terra-ui-packages/core-utils": "*",
```
```suggestion
    "@terra-ui-packages/core-utils": "^0.1.0",
```

For anything that's going to be published, let's use actual version specifiers instead of `*`. `*` is really only for the Terra UI application, which always uses the latest version in the repository.

nawatts commented on packages/notifications/src/error.test.ts:
```diff
@@ -0,0 +1,121 @@
+    let outerThrow: unknown | null = null;
+
+    // Act
+    try {
+      await callMe();
+    } catch (e) {
+      outerThrow = e;
+    }
```
Consider using [toThrow](https://jestjs.io/docs/expect#tothrowerror) to check for an expected exception.

```suggestion
    await expect(callMe).rejects.toThrow(expectedError);
```

msilva-broad replied:
I go back and forth on this... I think the current form reads better through the AAA sections... whereas using expect....toThrow technically has the Act in the Assert section....

nawatts replied:
🤷 I just write the comment as "Act / Assert". It's a lot more concise and avoids the mutable variable.

msilva-broad replied:
ok yeah, I'll switch them over.

nawatts commented on packages/notifications/src/error.test.ts:
```diff
@@ -0,0 +1,121 @@
+describe('reportError', () => {
+  it('calls notify and console.error', async () => {
+    // Arrange
+    jest.spyOn(console, 'error').mockImplementation(() => {});
```
Could this be moved to a `beforeEach` block instead of repeating in every test?

msilva-broad replied:
I err on the side of caution with this... since putting it in beforeEach would silence ALL tests console errors, which can be bad... so intentionally putting it in each test instead.

nawatts replied:
Fair enough. Though I'd claim that since this is the `reportError` describe block, I'll assume that every test in the block calls `reportError` and thus we'll want to silence console errors.

nawatts commented on packages/notifications/src/error.ts:
```diff
@@ -0,0 +1,87 @@
+ * allows composition of NotificationsContract dependency when useful for code reuse
+ * @param notifier
+ */
+export const makeNotificationsContract = (
+  notifier: Notifier,
+  ignoreError: IgnoreErrorDecider
+): NotificationsContract => ({
```
Can we turn the second parameter into an options object to clarify its purpose / make it optional?

```suggestion
interface NotificationsContractOptions {
  ignoreError?: IgnoreErrorDecider;
}

export const makeNotificationsContract = (
  notifier: Notifier,
  options: NotificationsContractOptions = {}
): NotificationsContract => {
  const { ignoreError = () => false } = options;
```

nawatts replied:
Also, maybe `shouldIgnoreError`? I think a `should...` prefix is a common convention for this sort of function (React's `shouldComponentUpdate`, etc).

nawatts replied:
I think my proposal would be an `ErrorReporter` interface (that does not include `notify`) and a function `makeNotificationErrorReporter` or something like that. Then the `NotificationsContract` returned by `useNotifications` could be a union type of `Notifier` and `ErrorReporter`.

msilva-broad replied:
right, but things are a bit entangled.... reportError method (which is the key functionality used by all the methods), needs to call .notify()... hence why it's all constructed together in the proposed implementation of makeNotificationsProvider

nawatts replied:
Yeah, I think this comment was resolved by consolidating everything into `NotificationsProvider` and thus avoiding "error reporting" as a thing.

nawatts commented on packages/notifications/src/error.ts:
```diff
@@ -0,0 +1,87 @@
+  reportErrorAndRethrow: (title: string) => {
+    return <F extends AnyPromiseFn>(fn: F) =>
+      withErrorHandling(async (error): Promise<void> => {
+        await makeNotificationsContract(notifier, ignoreError).reportError(title, error);
```
Instead of recursing, can we either declare all these methods individually so that they can call each other and return an object at the end or make notifications contract a class so methods can call other methods?

nawatts commented on packages/notifications/src/error.ts:
```diff
@@ -0,0 +1,87 @@
+      })(fn);
+  },
+
+  /**
+   *  This function is designed for use in modals.
+   *  Modals can overlay any error reporting, with the `throw` in the default `withErrorReporting`
+   *  preventing the modal itself from closing on error
+   *  As such, we must ensure we call the dismiss function if an error occurs
+   */
+  withErrorReportingInModal: (title: string, onDismiss: () => void) => {
+    return withErrorHandling(async (error) => {
+      await makeNotificationsContract(notifier, ignoreError).reportError(title, error);
+      onDismiss();
+      throw error;
+    });
+  },
```
Not too keen on concerning this package with modals. I think this could be defined in Terra UI.

Also, generally, I think this function encourages lazy error handling and would vote against promoting it to a package. Often a modal involves submitting some form and 400 errors should be handled by showing which field was invalid and allowing the user to adjust instead of closing the modal, etc.

msilva-broad replied:
I would argue that there are plenty of less complex modals where handing off to an "error notification" is fine.  Also, at least one subcomponent/dependency of Environments component uses withErrorReportingInModal, so having it be part of the provided contract from context works well. And generally, modals are a pretty common UX mechanic to accomodate...

nawatts replied:
I think I'd still vote for getting the "modal" specific names out of it. This can be more generic... it's the default error reporting plus some `onError` callback.

msilva-broad replied:
hmmm... yeah, the variants are a bit odd if you look a them collectively.... including reportErrorAndRethrow should probably be named withErrorReportingandRethrow since it returns a function agumenter.

what about simplifying like this:

```
export interface ErrorReportingOptions {
  rethrow?: boolean; // default false
  onNotified?: () => void;
}

export interface NotificationsContract extends Notifier {
  reportError: (title: string, obj?: unknown) => Promise<void>;

  /**
   * Returns a function augmenter (HoF).
   * If provided function fails when called, report the error to the user with `title` as a side effect.
   * Use options arg to give onNotified callback, and/or rethrow on error.
   */
  withErrorReporting: <F extends AnyPromiseFn>(title: string, options?: ErrorReportingOptions) => (fn: F) => F;
}
```

nawatts replied:
Sounds good. A few nits to avoid leaking "notifications" out of "error reporting"...

```ts
export interface ErrorReportingOptions {
  rethrow?: boolean; // default false
  onReport?: () => void;
}

export interface ErrorReporter {
  reportError: (title: string, obj?: unknown) => Promise<void>;

  /**
   * Returns a function augmenter (HoF).
   * If provided function fails when called, report the error to the user with `title` as a side effect.
   * Use options arg to give onNotified callback, and/or rethrow on error.
   */
  withErrorReporting: <F extends AnyPromiseFn>(title: string, options?: ErrorReportingOptions) => (fn: F) => F;
}
```

msilva-broad replied:
ok cool, tho this last point blends into your other comment.... the tricky bit is that Environments (and similar large components) have components and hooks that need both Notifier.notify and the methods we are settling on here for "ErrorReporter".. This PR is blending them together into the NotificationsContract so that we have just one "thing" to worry about in context.... trying to avoid context explosion... thoughts?

nawatts replied:
My thinking is that "error reporting" is a thing, then it's separate from (though can be built on) the base "notifications".
```ts
type NotificationsContract = Notifier & ErrorReporter;
```

Or, if "error reporting" is part of "notifications", then "error reporting" is not a thing on its own and does not appear in types.
```ts
type Notifications extends BaseNotifications {
  reportError: ...
  ...
}
```

nawatts commented on packages/notifications/src/error.ts:
```diff
@@ -0,0 +1,87 @@
+ * allows composition of NotificationsContract dependency when useful for code reuse
+ * @param notifier
+ */
+export const makeNotificationsContract = (
```
Not keen on this name... we've been using "contract" for types, not concrete objects.

nawatts commented on src/libs/error.test.ts:
```diff
@@ -1,114 +1,36 @@
 jest.mock('src/libs/notifications');
 
-describe('report', () => {
+describe('reportError - using terra notificationsProvider fallback', () => {
```
More hints that the name could use some refinement... here it's called a "notifications provider".

nawatts commented on src/libs/error.ts:
```diff
@@ -1,91 +1,24 @@
 
-// provide backwards compatible fail-safe for existing code that is not yet ready for code-reuse
-const withTerraUIReporter = withErrorReporter(notificationsProvider);
+export const terraUIReporter = makeNotificationsContract(notifier, ignoreSessionTimeout);
```
Another hint that the name could use some refinement. Here it's called a "reporter".

evrii commented on packages/notifications/src/error.test.ts:
```diff
@@ -0,0 +1,121 @@
+    const { reportError } = makeNotificationsContract(mockNotifications, ignoreError);
+
+    // Act
+    await reportError('Things went BOOM', new Error('BOOM!'));
```
Could the parameters be stored as variables and then used in your assertions?

msilva-broad replied:
I tend to intentionally avoid that to guard against accidental object === object assertions later in the test.  You can be careful to use the correct "deep equality" assertion, but this intentional duplication is just muscle memory and a 2nd failsafe I've gotten used to using.

evrii replied:
That makes sense. Thanks for explaining :-)

nawatts replied:
Although... referential equality implies value equality. The reverse is not true, so that's where you have to be careful.

evrii commented on packages/notifications/src/error.test.ts:
```diff
@@ -0,0 +1,121 @@
+    expect(console.error).toBeCalledWith('Things went BOOM', expectedError);
+  });
+
+  it('ignores session timeout error', async () => {
```
Where is the session timeout coming from? Is it worth adding a test confirming that the session timeout actually happens?

msilva-broad replied:
there's a test over in terra-ui app land that tests a part of this more specific to session timeout error check... here it should be renamed to something more generic to ignoring an error.

evrii commented on packages/notifications/src/error.test.ts:
```diff
@@ -0,0 +1,121 @@
+});
+
+describe('withErrorReporting', () => {
+  it('calls dismiss and notify', async () => {
```
How do you know that dismiss was called?

msilva-broad replied:
oops, that was a cut-and-pasted test title from the test below that does check dismiss.  I will make sure this gets fixed with the other PR feedback changes I am working on.

msilva-broad committed:
UIE-165 error.ts more decoupling for Environments code sharing
- PR feedback, naming tweaks.
- consolidated most error reporting methods into paramaterized withErrorReporting method.  Updated terra-ui error.ts fallbacks to forward to new method.

nawatts commented on packages/notifications/src/error.ts:
```diff
@@ -31,57 +31,37 @@ export const withErrorHandling = (callback: ErrorCallback) => {
-  ignoreError: IgnoreErrorDecider
-): NotificationsContract => ({
-  notify: notifier.notify,
+  shouldIgnoreError: IgnoreErrorDecider
```
Suggest making the second argument an options object so that it's self-documenting.

nawatts commented on packages/notifications/src/error.ts:
"error" is no longer a really appropriate name for this module since it's constructing the entire `NotificationsProvider` and not confined to errors specifically. Maybe this could be replaced by a `NotificationsProvider` module that includes the type definitions and `makeNotificationsProvider` function?

msilva-broad replied:
ok, shifting most things to a notifications-provider.ts module, but kept a small error.ts module for the couple of simple error methods.

nawatts commented on packages/notifications/src/error.ts:
```diff
@@ -0,0 +1,67 @@
+    withErrorReporting: (title: string, args?: ErrorReportingOptions) => {
+      return withErrorHandling(async (error) => {
+        await notifications.reportError(title, error);
+        if (args && args.onReported) {
+          args.onReported();
+        }
+        if (args && args.rethrow) {
+          throw error;
+        }
```
Nit: optional chaining would make this more concise.

```suggestion
        args?.onReported?.();
        if (args?.rethrow) {
          throw error;
        }
```

nawatts commented on packages/notifications/src/useNotifications.ts:
```diff
@@ -8,29 +9,53 @@ export interface NotificationOptions {
 };
-/** Gets the current NotificationsProvider from React context. */
-export const useNotificationsFromContext = (): NotificationsContract => {
+/** Gets the current NotificationsContextProvider from React context. */
```
This comment was correct before.

```suggestion
/** Gets the current NotificationsProvider from React context. */
```

msilva-broad committed:
UIE-165 error.ts more decoupling for Environments code sharing
- shifted most NotificationsProvider bits into notifications-provider.ts module.
- ts error fix and minor cleanup.

nawatts commented on packages/notifications/src/index.ts:
```diff
@@ -1 +1,3 @@
+export * from './error';
```
Should `withErrorHandling` and `withErrorIgnoring` be exported? They're more internals for `NotificationProvider`. On their own, they're not really notifications related.

Ideally... these might go in `core-utils`?

nawatts replied:
I think I'd vote not export them in this PR and then we can move them (and deal with the related version bump requirements) in a follow on?

msilva-broad replied:
a smaller followup PR would be good... but they have to be exported since they are re-exported in src/libs/error.ts for pretty wide use across terra-ui codebase.

nawatts replied:
Oh, I see.

nawatts replied:
Alternatively, libs/error could temporarily duplicate these functions until a follow up adds them to core-utils where both notifications and Terra UI can pull them from. But not much difference as long as we don't publish a version of notifications with these exports.

nawatts commented on packages/notifications/src/notifications-provider.ts:
```diff
@@ -0,0 +1,77 @@
+ */
+export const makeNotificationsProvider = (
+  notifier: Notifier,
+  shouldIgnoreError: IgnoreErrorDecider
```
I'd still vote for an options object here.

msilva-broad replied:
agreed... wanted to clear the deck and do that as the next commit.

msilva-broad committed:
UIE-165 error.ts more decoupling for Environments code sharing
- changed args to makeNotificationsProvider factory method to be a single args object so that usage is more self-documenting.

nawatts commented on packages/notifications/src/notifications-provider.ts:
```diff
@@ -0,0 +1,82 @@
+ * allows composition of NotificationsProvider dependency when useful for code reuse
+ * @param deps
+ */
+export const makeNotificationsProvider = (deps: NotificationsProviderDeps): NotificationsProvider => {
```
For the record, my suggestion was actually that only the second argument was an options object (because it contained _optional_ arguments).

`notifier` is always required, so it makes sense to be a positional argument.

Something like:

```suggestion
export const makeNotificationsProvider = (notifier: Notifier, options: NotificationsProviderOptions = {}): NotificationsProvider => {
  const { shouldIgnoreError = () => false } = options;
```

msilva-broad replied:
yeah, I know, but I do like expressive arg objects in general... also, both args are actually required, tho I thought about making the 2nd one optional.

jdcanas reviewed (approved):
looks good from IA perspective

msilva-broad committed:
UIE-165 error.ts more decoupling for Environments code sharing
- rebase

cahrens commented on src/workspaces/common/state/useWorkspaces.ts:
```diff
@@ -25,7 +24,7 @@ const defaultFieldsArgs: FieldsArg = [
 export const useWorkspaces = (fieldsArg?: FieldsArg, stringAttributeMaxLength?: number): UseWorkspacesResult => {
   const signal = useCancellation();
-  const { reportError } = withErrorReporter(useNotificationsFromContext());
+  const { reportError } = useNotificationsFromContext();
```
Why the removal of `withErrorReporter` here?

msilva-broad replied:
the core change that enabled full decoupling of errors.ts and notifications was to have the NotificationsProvider returned by useNotificationsFromContext to now include the reportError, withErrorReporting methods in its interface that previously had come through withErrorReporter helper.

the mechanics of what used to be withErrorReporter helper are now pre-backed into the constructed NotificationsProvider injected into react context at the root of app visual tree in Main.ts  There we use terraUINotifications export from what remains of src/libs/error.ts where the construction is done.

cahrens replied:
Ah, I see. I looked through more of the code and understand now.

cahrens commented on src/billing/List/BillingList.ts:
```diff
@@ -86,17 +87,17 @@ export const BillingList = (props: BillingListProps) => {
-        .Billing.listAccounts()
-        .then(_.keyBy('accountName')) // @ts-ignore
-        .then(setBillingAccounts);
+  const loadAccounts = withHandlers(
```
Was this change actually needed for this PR? Or are you just demonstrating this `withHandlers` function?

msilva-broad replied:
oh yeah, this was to fix a typescript error that was surfaced once one of the withXxxx handlers in _.flow had better types.

cahrens reviewed (commented):
Leaving comments… hopefully.
